### ISSUE TITLE
Sort dataset by loss

### ISSUE THREAD

**Issue by dmagas-auto1** [PyTorchLightning/pytorch-lightning#4091]
## 🚀 Feature
Ability to sort dataset by loss, can be in "Common use cases" or "Debugging" section

### Motivation
Inspired by this tweet:
https://twitter.com/karpathy/status/1311884485676294151

### Alternatives

What I do now: set batch size to 1, iterate over dataloader and caculate the loss for each example. This is really slow (probably because of batch size and my usage of detach() and item()) and I would like to speed it up, open to suggestions:
```
r = []
for batch in tqdm(val_dataloader):
    yhat = model(batch[0])
    y = batch[1]
    loss = calc_loss(yhat, y)
    r.append([batch[2][0], loss])
# batch[2] is a filepath
```

### Additional context

I can also try to make a PR with this feature but don't know where to start

**Comment by JackCaster** [comment]
I am also interested in this. I wonder if we could come up with a callback to keep track of the loss for each element in the batch (train, val, and test). For example, I am working on a classification task with a rather simple dataset. I can't seem to increase the accuracy despite many attempts. If I could sort the dataset by loss I am sure I would find the images that are mislabelled

**Comment by dmagas-auto1** [comment]
In the end I did this (I also have `images_files_paths` in the lightning module and used metric `nn.CrossEntropyLoss(reduction="none")` and manual mean in the training step)
```from pytorch_lightning.callbacks import Callback


class MyCallback(Callback):
    def __init__(self):
        self.losses = []
        self.batch_idxs = []
        self.val_losses = []

    def on_train_batch_end(
        self, trainer, pl_module, outputs, batch, batch_idx, dataloader_idx
    ):
        try:
            batch_loss = pl_module.batch_loss
        except AttributeError as err:
            m = """please track the batch_loss in the training_step like so:
                def training_step(...):
                    self.batch_loss = loss.sum(0).detach().tolist()
            """
            raise AttributeError(m) from err
        self.losses += batch_loss
        self.batch_idxs.append(batch_idx)
    
    def on_validation_batch_end(
        self, trainer, pl_module, outputs, batch, batch_idx, dataloader_idx
    ):
        try:
            batch_loss = pl_module.batch_loss
        except AttributeError as err:
            m = """please track the batch_loss in the training_step like so:
                def training_step(...):
                    self.batch_loss = loss.sum(0).detach().tolist()
            """
            raise AttributeError(m) from err
        self.val_losses += batch_loss

mcb = MyCallback() # and put it into trainer
# and after 1 epoch I do 

losses = mcb.losses[:len(model.images_files_paths)] 
df = pd.DataFrame({'loss': losses, 'path': model.images_files_paths})
